packages/cursorless-engine/src/actions/BringMoveSwap.ts

```diff
@@ -270,80 +253,157 @@ class BringMoveSwap implements Action {
+  constructor(rangeUpdater: RangeUpdater) {
+    super(rangeUpdater, "bring");
   }
 
-  async run([sources, destinations]: [
-    Target[],
-    Target[],
-  ]): Promise<ActionReturnValue> {
+  async run(
+    sources: Target[],
+    destinations: Destination[],
+  ): Promise<ActionReturnValue> {
     sources = this.broadcastSource(sources, destinations);
 
-    await this.decorateTargets(sources, destinations);
+    await this.decorateTargets(
+      sources,
+      destinations.map((d) => d.target),
+    );
 
-    const edits = this.getEdits(sources, destinations);
+    const edits = this.getEditsBringMove(sources, destinations);
 
     const markEntries = await this.performEditsAndComputeThatMark(edits);
 
-    const { thatMark, sourceMark } = this.calculateMarks(markEntries);
+    const { thatMark, sourceMark } = this.calculateMarksBringMove(markEntries);
 
     await this.decorateThatMark(thatMark);
 
     return { thatSelections: thatMark, sourceSelections: sourceMark };
   }
 }
 
-export class Bring extends BringMoveSwap {
-  constructor(rangeUpdater: RangeUpdater) {
-    super(rangeUpdater, "bring");
-  }
-}
-
 export class Move extends BringMoveSwap {
+  decoration = {
+    sourceStyle: FlashStyle.pendingDelete,
+    destinationStyle: FlashStyle.pendingModification0,
+    getSourceRangeCallback: getRemovalHighlightRange,
+  };
+
   constructor(rangeUpdater: RangeUpdater) {
     super(rangeUpdater, "move");
   }
+
+  async run(
+    sources: Target[],
+    destinations: Destination[],
+  ): Promise<ActionReturnValue> {
+    sources = this.broadcastSource(sources, destinations);
+
+    await this.decorateTargets(
+      sources,
+      destinations.map((d) => d.target),
+    );
+
+    const edits = this.getEditsBringMove(sources, destinations);
+
+    const markEntries = await this.performEditsAndComputeThatMark(edits);
+
+    const { thatMark, sourceMark } = this.calculateMarksBringMove(markEntries);
+
+    await this.decorateThatMark(thatMark);
+
+    return { thatSelections: thatMark, sourceSelections: sourceMark };
+  }
 }
 
 export class Swap extends BringMoveSwap {
+  decoration = {
+    sourceStyle: FlashStyle.pendingModification1,
+    destinationStyle: FlashStyle.pendingModification0,
+    getSourceRangeCallback: getContentRange,
+  };
+
   constructor(rangeUpdater: RangeUpdater) {
     super(rangeUpdater, "swap");
   }
+
+  async run(
+    targets1: Target[],
+    targets2: Target[],
+  ): Promise<ActionReturnValue> {
+    await this.decorateTargets(targets1, targets2);
+
+    const edits = this.getEditsSwap(targets1, targets2);
+
+    const markEntries = await this.performEditsAndComputeThatMark(edits);
+
+    await this.decorateThatMark(markEntries);
+
+    return { thatSelections: markEntries };
+  }
+
+  private getEditsSwap(targets1: Target[], targets2: Target[]): ExtendedEdit[] {
+    const results: ExtendedEdit[] = [];
+
+    targets1.forEach((source, i) => {
+      const destination = targets2[i];
```

Comment:
I was imagining you'd convert the target to a destination on the fly here when you use it as a destination